app/models/inventory.rb

```diff
@@ -7,11 +7,74 @@ class Inventory < ApplicationRecord
     inactive: 2
   }, _prefix: true
 
+  before_save :downcase_color
+
   validates_associated :model
   validates :memory_size, :manufactoring_year, :os_version, :color, :price, presence: true
 
   validates :memory_size, numericality: { only_integer: true, greater_than: 0 }
   validates :manufactoring_year, numericality: { only_integer: true, greater_than: 0,
                                                  less_than_or_equal_to: Time.zone.now.year }
   validates :price, numericality: { greater_than_or_equal_to: 0 }
+
+  # Search / Filter functions
+  scope :by_id, -> (id) { where id: id }
+  scope :by_model_id, -> (model_id) { where model_id: model_id }
+  scope :by_manufactoring_year_lb, -> (year) { where "manufactoring_year >= ?", year }
+  scope :by_manufactoring_year_ub, -> (year) { where "manufactoring_year <= ?", year }
+  scope :by_status, -> (status) { where status: status }
+  scope :by_os_version, -> (os) { where "os_version LIKE ?", "%" + os + "%" }
+  scope :by_color, -> (color) { where "color LIKE ?", "%" + color.downcase + "%" }
+
+  def self.search(query_params)
+    # return find(query_params[:id]) unless query_params[:id].blank?
```

Comment:
```suggestion
    # return find(query_params[:id]) unless query_params[:id].blank?
    return where(id: query_params[:id]) unless query_params[:id].blank?
```

Reply:
Should I handle the case that user inputs ID as well as other fields which don't match any record? As I understand your suggestion, we ignore other fields and return matched ID record anyway.

Reply:
It is your decision.
In my opinion, If I want to find exactly a product, I just only focus on id fields and others could be blank.